ingest/de.py

```diff
@@ -319,24 +331,63 @@ def all_match_ensembl_id_regex(adata, col_name):
         regex = r'ENS[A-Z]{0,4}\d{11}'
         return adata[col_name].astype(str).str.match(regex).all()
 
+    @staticmethod
+    def write_de_result(adata, group, annotation, rank_key, cluster_name, extra_params):
+        de_type =  extra_params.get("de_type")
+        method = extra_params.get("method")
+        annot_scope = extra_params.get("annotation_scope")
+        clean_annotation = DifferentialExpression.sanitize_string(annotation)
+        rank = sc.get.rank_genes_groups_df(adata, key=rank_key, group=group)
+        if DifferentialExpression.all_match_ensembl_id_regex(rank, "names"):
+            feature_name_rank = rank.merge(adata.var['feature_name'], left_on="names", right_index=True, how='left')
+            feature_name_rank.rename(columns={'names':'feature_id'}, inplace=True)
+            feature_name_rank.rename(columns={'feature_name':'names'}, inplace=True)
+            new_column_order = ["names"] + [col for col in feature_name_rank.columns if col not in {"names", "feature_id"}] + ["feature_id"]
+            feature_name_rank = feature_name_rank[new_column_order]
+
+            rank = feature_name_rank
+        if DifferentialExpression.delimiter_in_gene_name(rank):
+            DifferentialExpression.extract_gene_id_for_out_file(rank)
+        if de_type == "rest":
+            clean_group = DifferentialExpression.sanitize_string(group)
+            out_file = f'{cluster_name}--{clean_annotation}--{clean_group}--{annot_scope}--{method}.tsv'
+            DifferentialExpression.de_logger.info(f"Writing DE output for {clean_group} vs restq")
+        elif de_type == "pairwise":
+            # rank_genes_groups accepts a list. For SCP pairwise, should be a list with one item
+            # converting list to string for incorporation into result filename
+            group1 = DifferentialExpression.sanitize_string(''.join(extra_params["group1"]))
+            group2 = DifferentialExpression.sanitize_string(extra_params["group2"])
+            out_file = f'{cluster_name}--{clean_annotation}--{group1}--{group2}--{annot_scope}--{method}.tsv'
+            DifferentialExpression.de_logger.info(f"Writing DE output for {group1} vs {group2} pairwise")
+        else:
+            msg = f'Unknown de_type, {de_type}'
+            print(msg)
+            log_exception(
+                DifferentialExpression.dev_logger, DifferentialExpression.de_logger, msg
+            )
+            raise ValueError(msg)
+        # Round numbers to 4 significant digits while respecting fixed point
+        # and scientific notation (note: trailing zeros are removed)
+        rank.to_csv(out_file, sep='\t', float_format='%.4g')
```

Comment:
Is there a scientific / engineering reason we're more precise for DE values (rounded to 4 significant digits) than for expression values and numeric annotation values ([rounded to 3](https://github.com/search?q=repo%3Abroadinstitute%2Fscp-ingest-pipeline+round&type=code))?

Either way, I see rounded DE values to 4 is pre-existing, so this seems OK in terms of this PR.